deduction_rules.py

```diff
@@ -0,0 +1,121 @@
+        cells_to_check = [(row,col) for col in range(9)]
+        made_deduction |= func(cells_to_check)
+
+    for col in range(9):
+        cells_to_check = [(row,col) for row in range(9)]
+        made_deduction |= func(cells_to_check)
+
+    for sec in range(9):
+        cells_to_check = [local_to_global(sec,i,j) for i,j in product(range(3),range(3))]
+        made_deduction |= func(cells_to_check)
+    return made_deduction
+
+def _ban_numbers(sudoku, cell,numbers,rule, cells_used):
+    made_deduction = False
+    for number in numbers:
+        made_deduction |= sudoku.ban(cell[0],cell[1],number,rule,cells_used)
+    return made_deduction
+
+def _allowed_numbers(sudoku, cells):
+    """Returns a list where i-th element contains the allowed numbers in cells[i]"""
+    allowed_numbers = []
+    for cell in cells:
+        if sudoku.board[cell[0]][cell[1]] != 0:
+            allowed_numbers.append((sudoku.board[cell[0]][cell[1]],))
+        else:
+            allowed_numbers.append(tuple(sudoku.allowed[cell[0]][cell[1]].allowed()))
+    return allowed_numbers
+
+
+def nake_pair(sudoku):
+    """RULE: if v and w can be written only to two cells in this row/col/sec, then remove it from the other cells of the row/col/sec."""
+    
+    def search_for_nake_pairs(elems):
+        """Returns a list of 2-tuples, where every tuple contains the indeces of the same input"""
+        return [(i,j) for i in range(len(elems)) for j in range(i+1,len(elems)) if len(elems[i])==2 and elems[i]==elems[j]]
+    
+    def search_and_ban_in_subset(cells_to_check):
+        """Searches all nake-pairs in a subset of cells and bans these numbers from the other elements of subset."""
+        made_deduction = False
+        allowed_numbers = _allowed_numbers(sudoku,cells_to_check)
+        nake_pairs = search_for_nake_pairs(allowed_numbers)
+        for pair in nake_pairs:
+            current_naked_cell1 = cells_to_check[pair[0]]
+            current_naked_cell2 = cells_to_check[pair[1]]
+            deleted_numbers = sudoku.allowed[current_naked_cell1[0]][current_naked_cell1[1]].allowed()
+            cells_used = sudoku.allowed[current_naked_cell1[0]][current_naked_cell1[1]].notNones()+sudoku.allowed[current_naked_cell2[0]][current_naked_cell2[1]].notNones()
+            for cell in cells_to_check:
+                if cell not in (cells_to_check[pair[0]],cells_to_check[pair[1]]):
+                    made_deduction |= _ban_numbers(sudoku,cell,deleted_numbers,"nake_pair",cells_used)
+        return made_deduction
+
+    return _apply_for_nines(search_and_ban_in_subset)
+
+def hidden_pair(sudoku):
+    """RULE: if only v and w has number x and y then delete every number from v and w except x and y."""
+    def search_hidden_pairs(elems):
+        where_can_go = {i:[] for i in range(1,10)} # i-th number in which indeces of elems can go
+        for idx,elem in enumerate(elems):
+            for num in elem:
+                where_can_go[num].append(idx)
```

Comment:
It is more complicated now to pass the index of row, col, sec, which cells I am currently spectating. So if it won't cause any problem in the feature, I don't want to fix these four lines, because I need to edit for that at least 10 lines.